File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<int> dist(n, INT_MAX);
        dist[src] = 0;

        for (int i = 0; i < k + 1; ++i) {
            vector<int> update_dist(n, INT_MAX);

            for (const auto& flight : flights) {
                if ((dist[flight[1]] - dist[flight[0]]) > flight[2]) {
                    update_dist[flight[1]] = min(update_dist[flight[1]], dist[flight[0]] + flight[2]);
                } 
            }

            for (int j = 0; j < n; ++j) {
                if (update_dist[j] != INT_MAX) {
                    dist[j] = update_dist[j];
                }
            }
        }
        return dist[dst] == INT_MAX ? -1 : dist[dst];
    }
};

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        graph = [[] for _ in range(n)]
        dist = [float("inf") for _ in range(n)]
        dist[src] = 0

        for fr, to, price in flights:
            graph[fr].append((to, price))

        for t in range(k + 1):
            cur_dist = deepcopy(dist)
            for start in range(n):
                for nbr, price in graph[start]:
                    cost = cur_dist[start] + price 
                    if cost < dist[nbr]:
                        dist[nbr] = cost
        
        return dist[dst] if dist[dst] != float("inf") else -1

        